sim-lib/src/lib.rs

```diff
@@ -1022,64 +1082,71 @@ async fn produce_simulation_results(
     nodes: HashMap<PublicKey, Arc<Mutex<dyn LightningNode>>>,
     mut output_receiver: Receiver<SimulationOutput>,
     results: Sender<(Payment, PaymentResult)>,
-    shutdown: Listener,
-) {
-    log::debug!("Simulation results producer started.");
-
+    listener: Listener,
+) -> Result<(), SimulationError> {
     let mut set = tokio::task::JoinSet::new();
 
     loop {
         tokio::select! {
             biased;
-            _ = shutdown.clone() => break,
+            _ = listener.clone() => {
+                return Ok(())
+            },
             output = output_receiver.recv() => {
-                match output{
+                match output {
                     Some(simulation_output) => {
                         match simulation_output{
                             SimulationOutput::SendPaymentSuccess(payment) => {
-                                let source_node = nodes.get(&payment.source).unwrap().clone();
-                                set.spawn(track_payment_result(
-                                    source_node, results.clone(), payment, shutdown.clone(),
-                                ));
+                                if let Some(source_node) = nodes.get(&payment.source) {
+                                    set.spawn(track_payment_result(
+                                        source_node.clone(), results.clone(), payment, listener.clone()
+                                    ));
+                                } else {
+                                    return Err(SimulationError::MissingNodeError(format!("Source node with public key: {} unavailable.", payment.source)));
+                                }
                             },
                             SimulationOutput::SendPaymentFailure(payment, result) => {
-                                if results.clone().send((payment, result)).await.is_err() {
-                                    log::debug!("Could not send payment result.");
+                                if results.send((payment, result.clone())).await.is_err() {
+                                    return Err(SimulationError::MpscChannelError(
+                                        format!("Failed to send payment result: {result} for payment {:?} dispatched at {:?}.", payment.hash, payment.dispatch_time),
+                                    ));
                                 }
                             }
                         };
-
                     },
-                    None => {
-                        return
+                    None => return Ok(())
+                }
+            },
+            track_payment = set.join_next() => {
```

Comment:
nit: add a TODO explaining that we're not going to wait for all tasks to exit, just so we don't forget